File: JavaScript Runtime Samples/JSRT Win32 Hosting Samples/Legacy JSRT Samples/C++/Profiler.cpp
#include "stdafx.h"
#include <msopc.h>
#include <string>
#include <stack>
#include <queue>

using namespace std;

Profiler::Profiler(void)
{
	m_refCount = 1;
}

Profiler::~Profiler(void)
{
}

HRESULT Profiler::QueryInterface(REFIID riid, void **ppvObj)
{
	if (riid == IID_IUnknown)
	{
		*ppvObj = (IUnknown *) this;
	}
	else if (riid == IID_IActiveScriptProfilerCallback)
	{
		*ppvObj = (IActiveScriptProfilerCallback *) this;
	}
	else if (riid == IID_IActiveScriptProfilerCallback2)
	{
		*ppvObj = (IActiveScriptProfilerCallback2 *) this;
	}
	else
	{
		*ppvObj = NULL;
		return E_NOINTERFACE;
	}

	AddRef();
	return NOERROR;
}

ULONG Profiler::AddRef()
{
	return InterlockedIncrement(&m_refCount);
}

ULONG Profiler::Release()
{
	long lw;

	if (0 == (lw = InterlockedDecrement(&m_refCount)))
	{
		delete this;
	}
	return lw;
}

HRESULT Profiler::Initialize(DWORD dwContext)
{
	fwprintf(stdout, L"Profiler::Initialize: 0x%lx\n", dwContext);
	return S_OK;
}

HRESULT Profiler::Shutdown(HRESULT hrReason)
{
	fwprintf(stdout, L"Profiler::Shutdown: 0x%lx\n", hrReason);
	return S_OK;
}

HRESULT Profiler::ScriptCompiled(PROFILER_TOKEN scriptId, PROFILER_SCRIPT_TYPE type, IUnknown *pIDebugDocumentContext)
{
	fwprintf(stdout, L"Profiler::ScriptCompiled: 0x%lx, %u\n", scriptId, type);
	return S_OK;
}

HRESULT Profiler::FunctionCompiled(PROFILER_TOKEN functionId, PROFILER_TOKEN scriptId, const wchar_t *pwszFunctionName, const wchar_t *pwszFunctionNameHint, IUnknown *pIDebugDocumentContext)
{
	fwprintf(stdout, L"Profiler::FunctionCompiled: 0x%lx, 0x%lx, %s, %s\n", scriptId, functionId, pwszFunctionName, pwszFunctionNameHint);
	return S_OK;
}

HRESULT Profiler::OnFunctionEnter(PROFILER_TOKEN scriptId, PROFILER_TOKEN functionId)
{
	fwprintf(stdout, L"Profiler::OnFunctionEnter: 0x%lx, 0x%lx\n", scriptId, functionId);
	return S_OK;
}

HRESULT Profiler::OnFunctionExit(PROFILER_TOKEN scriptId, PROFILER_TOKEN functionId)
{
	fwprintf(stdout, L"Profiler::OnFunctionExit: 0x%lx, 0x%lx\n", scriptId, functionId);
	return S_OK;
}

HRESULT Profiler::OnFunctionEnterByName(const wchar_t *pwszFunctionName, PROFILER_SCRIPT_TYPE type)
{
	fwprintf(stdout, L"Profiler::OnFunctionEnterByName: %s, %u\n", pwszFunctionName, type);
	return S_OK;
}

HRESULT Profiler::OnFunctionExitByName(const wchar_t *pwszFunctionName, PROFILER_SCRIPT_TYPE type)
{
	fwprintf(stdout, L"Profiler::OnFunctionExitByName: %s, %u\n", pwszFunctionName, type);
	return S_OK;
}

File: JavaScript Runtime Samples/JSRT Win32 Hosting Samples/Legacy JSRT Samples/VB/Profiler.vb
Imports System.Runtime.InteropServices
Imports ChakraHost.Hosting

<Guid("C3C44476-A1CE-403E-AFE2-D53EF618B139")>
Friend Class Profiler
    Implements IActiveScriptProfilerCallback2

    Public Sub OnFunctionEnterByName(pwszFunctionName As String, type As ProfilerScriptType) Implements IActiveScriptProfilerCallback2.OnFunctionEnterByName
        Console.WriteLine("Profiler.OnFunctionEnterByName: {0}, {1}", pwszFunctionName, type)
    End Sub

    Public Sub OnFunctionExitByName(pwszFunctionName As String, type As ProfilerScriptType) Implements IActiveScriptProfilerCallback2.OnFunctionExitByName
        Console.WriteLine("Profiler.OnFunctionExitByName: {0}, {1}", pwszFunctionName, type)
    End Sub

    Public Sub Initialize(dwContext As UInteger) Implements IActiveScriptProfilerCallback.Initialize
        Console.WriteLine("Profiler.Initialize: {0}", dwContext)
    End Sub

    Public Sub Shutdown(hrReason As UInteger) Implements IActiveScriptProfilerCallback.Shutdown
        Console.WriteLine("Profiler.Shutdown: {0}", hrReason)
    End Sub

    Public Sub ScriptCompiled(scriptId As Integer, type As ProfilerScriptType, pIDebugDocumentContext As IntPtr) Implements IActiveScriptProfilerCallback.ScriptCompiled
        Console.WriteLine("Profiler.ScriptCompiled: {0}, {1}", scriptId, type)
    End Sub

    Public Sub FunctionCompiled(functionId As Integer, scriptId As Integer, pwszFunctionName As String, pwszFunctionNameHint As String, pIDebugDocumentContext As IntPtr) Implements IActiveScriptProfilerCallback.FunctionCompiled
        Console.WriteLine("Profiler.FunctionCompiled: {0}, {1}, {2}, {3}", functionId, scriptId, pwszFunctionName, pwszFunctionNameHint)
    End Sub

    Public Sub OnFunctionEnter(scriptId As Integer, functionId As Integer) Implements IActiveScriptProfilerCallback.OnFunctionEnter
        Console.WriteLine("Profiler.OnFunctionEnter: {0}, {1}", scriptId, functionId)
    End Sub

    Public Sub OnFunctionExit(scriptId As Integer, functionId As Integer) Implements IActiveScriptProfilerCallback.OnFunctionExit
        Console.WriteLine("Profiler.OnFunctionExit: {0}, {1}", scriptId, functionId)
    End Sub
End Class
